aten/src/ATen/native/cpu/SortingKernel.cpp

```diff
@@ -83,6 +93,104 @@ struct KeyValueCompDesc {
   }
 };
 
+template <typename scalar_t>
+std::pair<scalar_t, scalar_t> get_min_max(const scalar_t* data, int64_t size) {
+  using Vec = vec::Vectorized<vec::vec_scalar_t<scalar_t>>;
+  int num_threads = at::get_num_threads();
+  std::vector<scalar_t> min_per_thread(num_threads, std::numeric_limits<scalar_t>::max());
+  std::vector<scalar_t> max_per_thread(num_threads, std::numeric_limits<scalar_t>::lowest());
+
+  at::parallel_for(0, size, at::internal::GRAIN_SIZE / num_threads, [&](int64_t begin, int64_t end) {
+    int tid = at::get_thread_num();
+    const scalar_t* local_data = data + begin;
+    const int64_t local_size = end - begin;
+    const auto local_min_max = at::vec::reduce2_all(
+            [](Vec x, Vec y) { return at::vec::minimum(x, y); },
+            [](Vec x, Vec y) { return at::vec::maximum(x, y); },
+            local_data,
+            local_size);
+    min_per_thread[tid] = local_min_max.first;
+    max_per_thread[tid] = local_min_max.second;
+  });
+
+  scalar_t global_min = min_per_thread[0];
+  scalar_t global_max = max_per_thread[0];
+  for (const auto i : c10::irange(1, num_threads)) {
+    if (min_per_thread[i] < global_min) {
+      global_min = min_per_thread[i];
+    }
+    if (max_per_thread[i] > global_max) {
+      global_max = max_per_thread[i];
+    }
+  }
+
+  return std::pair<scalar_t, scalar_t>(global_min, global_max);
+}
+
+static void parallel_sort1d_kernel(
+    const TensorBase& values,
+    const TensorBase& indices,
+    bool descending) {
+  // this kernel does not care about `stable` parameter as radix sort
+  // used here is a stable sorting algorithm
+  AT_DISPATCH_INTEGRAL_TYPES(values.scalar_type(), "parallel_sort1d_kernel", [&] {
+    scalar_t min = std::numeric_limits<scalar_t>::max();
+    scalar_t max = std::numeric_limits<scalar_t>::lowest();
+    std::tie(min, max) =
+        get_min_max(values.data_ptr<scalar_t>(), values.size(0));
+
+    int64_t elements = values.numel();
+    scalar_t* keys = values.data_ptr<scalar_t>();
+    int64_t* vals = indices.data_ptr<int64_t>();
+    std::vector<scalar_t> tmp_keys(elements);
+    std::vector<int64_t> tmp_vals(elements);
+    scalar_t* sorted_keys = nullptr;
+    int64_t* sorted_vals = nullptr;
+    std::tie(sorted_keys, sorted_vals) = radix_sort_parallel(
+        keys,
+        vals,
+        tmp_keys.data(),
+        tmp_vals.data(),
+        elements,
+        std::max(std::abs(min), std::abs(max)));
+
+    const bool sorted_in_place = keys == sorted_keys;
+    if (!sorted_in_place) {
+      std::memcpy(keys, sorted_keys, values.numel() * values.itemsize());
+      std::memcpy(vals, sorted_vals, indices.numel() * indices.itemsize());
```

Comment:
also if we need to try to fuse multiple omp sessions (if possible) into one so as to minimize the omp thread sync overhead.

for example, we can have `vals` and `keys` into one.